### PR TITLE
ar71xx: add support for COMFAST E355AC

### PR DESCRIPTION
**Pull request by HenryGiraldo**
Add support for COMFAST CF-E355AC 1200Mbps Gigabit WI-FI Router.

Initialization on mach-cf-e316n-v2.c but I think it's better to create a mach-comfast instead of using cf-e316n-v2 which is a specific hardware.

### CODE REVIEW

**Comment by pepe2k** [comment]
Hello @HenryGiraldo,

Please, read our [submitting patches rules](https://lede-project.org/submitting-patches), and make sure your patch or pull request follows them. Your commit is missing at least a description and SoB line.

Cheers,
Piotr

File: target/linux/ar71xx/files/arch/mips/ath79/mach-cf-e316n-v2.c [hunk 1]
```diff
@@ -130,6 +130,58 @@ static struct gpio_keys_button cf_e320n_v2_gpio_keys[] __initdata = {
...
+#define CF_E355AC_GPIO_LED_LAN          3
+#define CF_E355AC_GPIO_LED_WLAN         0
+
+#define CF_E355AC_GPIO_LED_EXT_WAN              16
+#define CF_E355AC_GPIO_LED_EXT_LAN              14
+#define CF_E355AC_GPIO_LED_EXT_WLAN             11
+
+static struct gpio_led cf_e355ac_leds_gpio[] __initdata = {
+        {
+                .name           = "comfast:red:wan",
```
Comment (heinzek) [inline target/linux/ar71xx/files/arch/mips/ath79/mach-cf-e316n-v2.c]: Use board name instead vendor in led name

File: target/linux/ar71xx/files/arch/mips/ath79/mach-cf-e316n-v2.c [hunk 2]
```diff
@@ -355,6 +407,42 @@ static void __init cf_e320n_v2_setup(void)
...
+        ath79_register_leds_gpio(-1, ARRAY_SIZE(cf_e355ac_leds_gpio),
+                                 cf_e355ac_leds_gpio);
+
+        ath79_register_gpio_keys_polled(1, CF_EXXXN_KEYS_POLL_INTERVAL,
+                                        ARRAY_SIZE(cf_e355ac_gpio_keys),
+                                        cf_e355ac_gpio_keys);
+}
+MIPS_MACHINE(ATH79_MACH_CF_E355AC, "CF-E355AC", "COMFAST CF-E355AC",
+             cf_e355ac_setup);
+
```
Comment (heinzek) [inline target/linux/ar71xx/files/arch/mips/ath79/mach-cf-e316n-v2.c]: use tab instead spaces in this file

File: target/linux/ar71xx/base-files/etc/board.d/01_leds [hunk 3]
```diff
@@ -181,6 +181,9 @@ cf-e320n-v2)
 	ucidef_set_led_netdev "wan" "WAN" "$board:red:wan" "eth1"
 	ucidef_set_led_wlan "wlan" "WLAN" "$board:blue:wlan" "phy0tpt"
 	;;
+cf-e355ac)
+	ucidef_set_led_netdev "lan" "LAN" "comfast:green:lan" "br-lan"
```
Comment (heinzek) [inline target/linux/ar71xx/base-files/etc/board.d/01_leds]: Change led name to $board:...:...

File: target/linux/ar71xx/files/arch/mips/ath79/Kconfig.openwrt [hunk 4]
```diff
@@ -1998,6 +1998,17 @@ config ATH79_MACH_CF_E320N_V2
 	select ATH79_DEV_USB
 	select ATH79_DEV_WMAC
 
+config ATH79_MACH_COMFAST_CF_E355AC
```
Comment (heinzek) [inline target/linux/ar71xx/files/arch/mips/ath79/Kconfig.openwrt]: Use boardname instead vendor_boardname

**Comment by HenryGiraldo** [comment]
@heinzek I have edited my commit with your suggestions.

Cheers,
He

File: target/linux/ar71xx/base-files/lib/ar71xx.sh [hunk 5]
```diff
@@ -504,6 +504,9 @@ ar71xx_board_detect() {
 	*"CF-E320N v2")
 		name="cf-e320n-v2"
 		;;
+	*"CF-E355AC")
```
Comment (pepe2k) [inline target/linux/ar71xx/base-files/lib/ar71xx.sh]: Are you sure that there is only one version of this device? As far as I can see on vendor website, v2 is mentioned in [firmware download section](http://en.comfast.com.cn/download/FirmwareUpgrade/).

File: target/linux/ar71xx/files/arch/mips/ath79/Kconfig.openwrt [hunk 6]
```diff
@@ -1998,6 +1998,15 @@ config ATH79_MACH_CF_E320N_V2
...
 
+config ATH79_MACH_CF_E355AC
+      bool "COMFAST CF-E355AC board support"
+      select SOC_QCA953X
+      select ATH79_DEV_ETH
+      select ATH79_DEV_GPIO_BUTTONS
+      select ATH79_DEV_LEDS_GPIO
+      select ATH79_DEV_M25P80
+
+
```
Comment (pepe2k) [inline target/linux/ar71xx/files/arch/mips/ath79/Kconfig.openwrt]: Drop this redundant empty line, please.

Also, your are missing here:
- `select ATH79_DEV_AP9X_PCI if PCI`
- `select ATH79_DEV_AP9X_PCI if PCI`

File: target/linux/ar71xx/files/arch/mips/ath79/mach-cf-e316n-v2.c [hunk 7]
```diff
@@ -130,6 +131,58 @@ static struct gpio_keys_button cf_e320n_v2_gpio_keys[] __initdata = {
...
+	}, {
+		.name           = "cf-e355ac:green:lan",
+		.gpio           = CF_E355AC_GPIO_LED_LAN,
+		.active_low     = 0,
+	}, {
+		.name           = "cf-e355ac:blue:wlan",
+		.gpio           = CF_E355AC_GPIO_LED_WLAN,
+		.active_low     = 0,
+	}, {
+		.name           = "cf-e355ac:blue:ext_wan",
```
Comment (pepe2k) [inline target/linux/ar71xx/files/arch/mips/ath79/mach-cf-e316n-v2.c]: Could you please explain what are all that `ext_xxx` LEDs?

File: target/linux/ar71xx/files/arch/mips/ath79/mach-cf-e316n-v2.c [hunk 8]
```diff
@@ -355,6 +408,43 @@ static void __init cf_e320n_v2_setup(void)
...
+static void __init cf_e355ac_common_setup(void)
+{
+	u8 *art = (u8 *) KSEG1ADDR(0x1f011000);
+	/* Disable JTAG, enabling GPIOs 0-3 */
+	/* Configure OBS4 line, for GPIO 4*/
+	ath79_gpio_function_setup(AR934X_GPIO_FUNC_JTAG_DISABLE, 0);
+
+	cf_exxxn_common_setup(0x10000, CF_E355AC_GPIO_XWDT_TRIGGER);
+
+	cf_exxxn_qca953x_eth_setup();
```
Comment (pepe2k) [inline target/linux/ar71xx/files/arch/mips/ath79/mach-cf-e316n-v2.c]: As you wrote in commit message, your board has only one Ethernet interface available, so you shouldn't use `cf_exxxn_qca953x_eth_setup()` as it initializes both interfaces.

File: target/linux/ar71xx/files/arch/mips/ath79/mach-cf-e316n-v2.c [hunk 9]
```diff
@@ -355,6 +408,43 @@ static void __init cf_e320n_v2_setup(void)
 MIPS_MACHINE(ATH79_MACH_CF_E320N_V2, "CF-E320N-V2", "COMFAST CF-E320N v2",
 	     cf_e320n_v2_setup);
 
+static void __init cf_e355ac_common_setup(void)
```
Comment (pepe2k) [inline target/linux/ar71xx/files/arch/mips/ath79/mach-cf-e316n-v2.c]: Please, drop this function.
Other `*_common_setup()` functions are called by more than one board init functions.

File: target/linux/ar71xx/mikrotik/config-default [hunk 10]
```diff
@@ -33,6 +33,7 @@
 # CONFIG_ATH79_MACH_CARAMBOLA2 is not set
 # CONFIG_ATH79_MACH_CF_E316N_V2 is not set
 # CONFIG_ATH79_MACH_CF_E320N_V2 is not set
+# CONFIG_ATH79_MACH_CF_E355AC is not set
```
Comment (pepe2k) [inline target/linux/ar71xx/mikrotik/config-default]: Please, do the same in `target/linux/ar71xx/nand/config-default`.

File: target/linux/ar71xx/base-files/lib/ar71xx.sh (hunk 5 above)
Comment (HenryGiraldo) [inline target/linux/ar71xx/base-files/lib/ar71xx.sh]: I have contacted the manufacturer and confirmed that there is only one version.

File: target/linux/ar71xx/files/arch/mips/ath79/mach-cf-e316n-v2.c (hunk 8 above)
Comment (HenryGiraldo) [inline target/linux/ar71xx/files/arch/mips/ath79/mach-cf-e316n-v2.c]: It has 2 ports, I gonna edit my board description message.

File: target/linux/ar71xx/base-files/lib/ar71xx.sh (hunk 5 above)
Comment (pepe2k) [inline target/linux/ar71xx/base-files/lib/ar71xx.sh]: OK, thank you for confirmation!

Cheers,
Piotr

File: target/linux/ar71xx/files/arch/mips/ath79/mach-cf-e316n-v2.c (hunk 7 above)
Comment (HenryGiraldo) [inline target/linux/ar71xx/files/arch/mips/ath79/mach-cf-e316n-v2.c]: I do not know exactly, I get from the manufacturer, when I clarify, I'll create a new commit.

Comment (pepe2k) [inline target/linux/ar71xx/files/arch/mips/ath79/mach-cf-e316n-v2.c]: > I do not know exactly, I get from the manufacturer, when I clarify, I'll create a new commit.

Are these LEDs available at all? Have you opened the device to see how many LEDs are actually on the board?

File: target/linux/ar71xx/base-files/etc/board.d/01_leds [hunk 11]
```diff
@@ -181,6 +181,9 @@ cf-e320n-v2)
 	ucidef_set_led_netdev "wan" "WAN" "$board:red:wan" "eth1"
 	ucidef_set_led_wlan "wlan" "WLAN" "$board:blue:wlan" "phy0tpt"
 	;;
+cf-e355ac)
+	ucidef_set_led_netdev "lan" "LAN" "$board:green:lan" "eth0"
```
Comment (pepe2k) [inline target/linux/ar71xx/base-files/etc/board.d/01_leds]: Don't duplicate code and move this line down, please (include it with other COMFAST boards).

File: target/linux/ar71xx/base-files/etc/board.d/01_leds [hunk 12]
```diff
@@ -182,7 +182,8 @@ cf-e320n-v2)
 	ucidef_set_led_wlan "wlan" "WLAN" "$board:blue:wlan" "phy0tpt"
 	;;
 cf-e380ac-v1|\
-cf-e380ac-v2)
+cf-e380ac-v2)\
+cf-e355ac)
 	ucidef_set_led_netdev "lan" "LAN" "$board:green:lan" "eth0"
 	ucidef_set_led_wlan "wlan2g" "WLAN2G" "$board:blue:wlan2g" "phy1tpt"
```
Comment (pepe2k) [inline target/linux/ar71xx/base-files/etc/board.d/01_leds]: Your board doesn't define `wlan2g` and `wlan5g` LEDs.
Are you testing your changes in code on device at all?

Comment (HenryGiraldo) [inline target/linux/ar71xx/base-files/etc/board.d/01_leds]: > Don't duplicate code and move this line down, please (include it with other COMFAST boards).
That's why I did this.

Comment (pepe2k) [inline target/linux/ar71xx/base-files/etc/board.d/01_leds]: My mistake, I wanted to write "move this line **up**"... but you could at least first check what LEDs your board has defined in mach-*.c file (in fact, you are the one who defined LEDs, so you should know what's there) and then include it in the correct place - you are here the one adding support for this board, I have never seen it.

Next time, if I write something stupid or make a typo in comment, just correct me.

As you can see:

```
cf-e320n-v2)
	ucidef_set_led_netdev "lan" "LAN" "$board:green:lan" "eth0"
	ucidef_set_led_netdev "wan" "WAN" "$board:red:wan" "eth1"
	ucidef_set_led_wlan "wlan" "WLAN" "$board:blue:wlan" "phy0tpt"
```
And your board has:

```
cf-e355ac:red:wan
cf-e355ac:green:lan
cf-e355ac:blue:wlan
```

Doesn't that look familiar for you? Of course, you should first make sure it's correct (lan/wan interfaces, radio - you have two radios on your board, maybe there are also two LEDs for both radios? The phy0tpt will be probably 5G radio, maybe you should then rename the LED).

Again, are you testing this code on a real device?

Comment (HenryGiraldo) [inline target/linux/ar71xx/base-files/etc/board.d/01_leds]: Yes, I'm testing the code, but with the LEDs there is a problem. It is a single RGB led that changes color, so it is difficult to see if it works well.

Comment (pepe2k) [inline target/linux/ar71xx/base-files/etc/board.d/01_leds]: > Yes, I'm testing the code, but with the LEDs there is a problem. It is a single RGB led that changes color, so it is difficult to see if it works well.

You can manipulate with LEDs manually, for example `echo 0 > /sys/class/leds/...led-name-here.../brightness`. With this you can manually shut down two LEDs and check if 3rd one works correctly (trigger, active low/high, etc.), then reload LEDs setting with `/etc/init.d/led reload` and repeat for other two LEDs.

Comment (HenryGiraldo) [inline target/linux/ar71xx/base-files/etc/board.d/01_leds]: I will redefine the use of led defined in mach-*.c file. Thank you very much for your advice.

**Comment by HenryGiraldo** [comment]
All requested changes are done, thx for your support @pepe2k

**Comment by HenryGiraldo** [comment]
Sorry, but I can not see what requested changes I have left to do.

**Comment by pepe2k** [comment]
@HenryGiraldo,

Thank you for your contribution! I have merged your changes (with minor fixes) into [my staging tree](https://git.lede-project.org/?p=lede/pepe2k/staging.git;a=summary).

Cheers,
Piotr

File: target/linux/ar71xx/base-files/lib/ar71xx.sh (hunk 5 above)
Comment (chrisgilpin) [reply target/linux/ar71xx/base-files/lib/ar71xx.sh]: Hi all, late to the game but there is indeed a v2 of this AP with different QC chips (I have the Access Point at my house now and am trying to get LEDE working). http://en.comfast.com.cn/product/WirelessAP/item-218.html

Comment (HenryGiraldo) [reply target/linux/ar71xx/base-files/lib/ar71xx.sh]: Is it a new model with Wave2? 
the manufacturer assured me that there was only one version of that model.

Comment (chrisgilpin) [reply target/linux/ar71xx/base-files/lib/ar71xx.sh]: Yes, according to http://en.comfast.com.cn/product/WirelessAP/item-218.html it has Wave2

Comment (pepe2k) [reply target/linux/ar71xx/base-files/lib/ar71xx.sh]: I'm so not surprised about v2 :)

**Pull request closed without merging**